primitives/src/circuit/merkle_tree.rs

```diff
@@ -10,26 +10,32 @@
 
 use crate::{
     circuit::rescue::RescueGadget,
-    merkle_tree::{AccMemberWitness, MerklePath, MerkleTree, NodePos, NodeValue},
+    merkle_tree::{
+        examples::RescueMerkleTree, internal::MerkleNode, MerkleTreeScheme, ToTraversalPath,
+    },
     rescue::RescueParameter,
 };
 use ark_ec::TEModelParameters as Parameters;
 use ark_ff::PrimeField;
-use ark_std::{vec, vec::Vec};
+use ark_std::vec::Vec;
 use jf_relation::{errors::CircuitError, BoolVar, Circuit, PlonkCircuit, Variable};
 
-#[derive(Clone)]
-struct MerkleNodeBooleanEncoding<F: PrimeField> {
-    sibling1: NodeValue<F>,
-    sibling2: NodeValue<F>,
+type NodeVal<F> = <RescueMerkleTree<F> as MerkleTreeScheme>::NodeValue;
+type MerkleProof<F> = <RescueMerkleTree<F> as MerkleTreeScheme>::MembershipProof;
+use typenum::U3;
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+struct MerkleNodeBooleanEncoding<F: PrimeField + RescueParameter> {
```

Comment:
Any reason that we need to add RescueParameter bound here?

Reply:
As I mentioned in the comment above, the current circuit is very specific for the rate-3 append-only MT instantiated with the Rescue hash, as specified in the #131 issue:
"Gadgets for append-only Merkle tree instantiated by rate-3 rescue hash function."
I'm using a concrete type from the `examples` module, whose trait bounds require `RescueParameters` on `F`.

As also hinted at in the above comment, I'd be in favour of making it very explicit that the circuit is for a specific type only. We could even move it somewhere else, maybe a separate module where we define both `pub type RescueMerkleTree<F> = MerkleTree<F, RescueHash<F>, u64, U3, F>;` and circuit impl.

Reply:
Yeah sounds good

Reply:
Done in https://github.com/EspressoSystems/jellyfish/pull/138/commits/4dcbd0b8a969fb3de588690c70408165ce435c50 and https://github.com/EspressoSystems/jellyfish/pull/138/commits/b61fa4e8213813f0876159bc075764a02ece1440.